### ISSUE TITLE
proposal: testing: add a flag to detect unnecessary skips

### ISSUE THREAD

**Issue by bcmills** [golang/go#25951]
**Motivation**

We often use `(*testing.T).Skip` to skip tests that fail due to known bugs in the Go toolchain, the standard library, or a specific platform where the test runs.

When the underlying bugs are fixed, it is important that we remove the skips to prevent regressions.

However, bugs are sometimes fixed without us noticing — especially if they are platform bugs. In such cases, it would be helpful to have some way to detect that situation short of editing and recompiling all of the tests in a project.

[Here](https://github.com/golang/go/blob/964639cc338db650ccadeafb7424bc8ebb2c0f6c/misc/cgo/test/issue17065.go#L24) is a concrete example in the standard library: a `t.Skip` call that references an issue that was marked closed back in 2016 (#17065). Given how easy that call was to find, I expect that similarly outdated skips are pervasive.

**Proposal**

A new boolean flag for the `testing` package, tentatively named `-test.unskip`, with the corresponding `go test` flag `-unskip`. `-test.unskip` takes a regular expression, which is matched against the formatted arguments of `t.Skip` or `t.Skipf` or the most recent test log entry before `t.SkipNow`.

If a `Skip` matches the regular expression, the test does not stop its execution as usual: instead, it continues to run. **If the skipped test passes, the test is recorded as unnecessarily-skipped and the binary will exit with a nonzero exit code.** If the skipped test fails, the test log (and the exit code of the test binary) ignores the failure and proceeds as usual.

**Comparison**

Test frameworks in a few other languages provide an explicit mechanism for expecting a test to fail. For example, Python has [`@unittest.expectedFailure`](http://go/pylib/unittest.html#skipping-tests-and-expected-failures); Boost has an [`expected_failures`](https://www.boost.org/doc/libs/1_67_0/libs/test/doc/html/boost_test/testing_tools/expected_failures.html#l_expected_failure) decorator; RSpec has the [pending](https://relishapp.com/rspec/rspec-core/v/3-7/docs/pending-and-skipped-examples) keyword.

**Comment by ianlancetaylor** [comment]
It's an interesting idea but I note that some tests are skipped not because they fail but because they impose excessive resource costs on some systems (beyond what `-test.short` blocks) or because they cause some systems to crash.  So this will push us in the direction of adding additional ways to skip a test.

**Comment by bcmills** [comment]
> some tests are skipped not because they fail but because they impose excessive resource costs on some systems (beyond what `-test.short` blocks) or because they cause some systems to crash.

That sort of use-case is part of the reason I proposed a regular expression rather than a boolean flag.

However, it certainly remains a usability consideration: for example, we wouldn't want to encourage people to run `-unskip=.` on unfamiliar low-level code (or on machines that can't tolerate a crash or accidental forkbomb).

**Comment by ianlancetaylor** [comment]
It would also be nice to handle flaky tests somehow.  Some tests are skipped because they are flaky on some platforms.  So the fact that they pass once doesn't indicate much one way or another.

For non-flaky tests I wonder if test.XFail("reason") would suffice to address this problem.  It would run the test as usual but flip the meaning of the result: a failure is a pass, a pass is a failure.  You mentioned this possibility in the original proposal; why not use it for Go?

**Comment by mvdan** [comment]
I agree that having an "expected failure" mode would be helpful.

A potentially related issue I've encountered in the past is making sure that CI doesn't skip tests when it shouldn't. For example, suppose that you write your tests to be skipped if a SQL database isn't running, and you set up one as part of CI. There currently isn't a way to tell `go test` never to skip those tests, other than doing `go test -v` and manually checking the `SKIP` output lines from time to time.

**Comment by bcmills** [comment]
> For non-flaky tests I wonder if test.XFail("reason") would suffice to address this problem.

That would probably suffice going forward, but introduces a migration problem: existing tests already use `Skip`.

That said, migrating `Skip`s to `XFail`s is a one-time cost. I'd be happy enough with it, I suspect.

**Comment by rsc** [comment]
XFail sounds interesting on its own right, but it still doesn't handle Skip-for-flaky. Let's put this on hold and circle back to package testing at some point later, during the Go 2 library phase.

**Comment by 13rac1** [comment]
I had the same feature request and decided to implement it as a learning exercise and potential contribution. Includes tests, but needs additional documentation: https://github.com/13rac1/go/commit/8536d923cd2aa658cf7563bec7bc50cd73cf56f8

Passing Test:
```
func TestFail(t *testing.T) {
	t.ExpectFail()
	t.Fail()
}

func TestFailNow(t *testing.T) {
	t.ExpectFail()
	t.FailNow()
}

func TestPanic(t *testing.T) {
	t.ExpectFail()
	panic("test")
}
```
Passing Results:
```
=== RUN   TestFail
--- PASS: TestFail (0.00s)
    testing.go:828: expected failure; test failed
=== RUN   TestFailNow
--- PASS: TestFailNow (0.00s)
    testing.go:828: expected failure; test failed
=== RUN   TestPanic
--- PASS: TestPanic (0.00s)
    testing.go:817: panic: test
    testing.go:828: expected failure; test failed
PASS
```
Failing Test:
```
func TestPass(t *testing.T) {
	t.ExpectFail()
	t.Log("pass")
}
```
Failing Results:
```
=== RUN   TestPass
--- FAIL: TestPass (0.00s)
    main_test.go:7: pass
    testing.go:831: expected failure; test did not fail
FAIL
```

**Comment by bcmills** [comment]
Here's a neat example:
https://github.com/golang/go/blob/a813d3c788b4ec58032616e8d269ee65d1b10085/src/cmd/go/go_test.go#L4692-L4694

Added in [CL 42533](https://golang.org/cl/42533) (May 4, 2017), rendered obsolete in [CL 46004](https://golang.org/cl/46004) (June 16, 2017), and skipping the test needlessly on an entire OS ever since.

**Comment by stdedos** [comment]
While this code may not satisfy the original purpose stated here, it's still very valuable for everyone coming from pytest testing methodologies.

tl;dr you are able to merge tests as they come, and if your CI is based on "everything passes or go home" it is not the end of the world; neither does code rot away in some personal repo, inactive.

**Comment by marco-m** [comment]
@rsc @bcmills @ianlancetaylor 

> XFail sounds interesting on its own right, but it still doesn't handle Skip-for-flaky. Let's put this on hold and circle back to package testing at some point later, during the Go 2 library phase.

The decision to put this proposal on hold is from June 2018; would it be possible to reconsider it now, since in my (limited) understanding Go 2 is still undecided?

My 2 cents is that also a simple XFail, that doesn't cover the case for skip-for-flaky, is still very useful in its own right, as the examples from @bcmills show.

Another advantage of a simple XFail is that it is "always on": no need to run the tests with a special flag "just in case something has been fixed and we should remove the Skip". As soon as a test marked XFail returns successfully, it will be reported as unexpected success.

Last advantage of a simple XFail is its immediate understandability by anybody seeing it for the first time.

**Comment by treuherz** [comment]
Having something like XFail wiuld also be helfpul for when building, or building on top of, testing helpers and frameworks. For example, I'm developing a custom matcher for [gomock](https://github.com/golang/mock), and like a test ensuring that certain inputs cause the test to fail. Currently there's no way I can see to do that.

The bare-bones approach of `testing` has made it easy for other projects to build tools and helpers for common cases on top of the library, and they typically take a very open approach to extensibility. Being able to develop or extend tools like this without being able to make assertions about how they'll interact with `testing` itself makes this much trickier.

**Comment by AlekSi** [comment]
See also #39903

**Comment by AlekSi** [comment]
We used something similar in FerretDB for several years. I finally extracted it into a simple self-contained library with no dependencies: https://github.com/AlekSi/xfail
